src/overlays/actors/ovl_En_Giant/z_en_giant.c

```diff
@@ -26,42 +31,449 @@ const ActorInit En_Giant_InitVars = {
+            }
+            break;
+        case 14:
+            this->alpha -= 12;
+            break;
+        default:
+            if (this->alpha < 255) {
+                this->alpha += 8;
+            }
+            break;
+    }
+}
+
+void EnGiant_PlayAndUpdateAnimation(EnGiant* this) {
+    if (SkelAnime_FrameUpdateMatrix(&this->skelAnime) &&
+        (this->animationId != GIANT_ANIMATION_FALLING_OVER || this->action != 6)) {
+        EnGiant_ChangeAnimation(this, this->animationId);
+        switch (this->action) {
+            case 3:
+                EnGiant_ChangeToStartOrLoopAnimation(this, GIANT_ANIMATION_LOOK_UP_START);
+                break;
+            case 4:
+                EnGiant_ChangeToStartOrLoopAnimation(this, GIANT_ANIMATION_RAISED_ARMS_START);
+                break;
+            case 5:
+                EnGiant_ChangeToStartOrLoopAnimation(this, GIANT_ANIMATION_STRUGGLE_START);
+                break;
+            case 6:
+                // Unused
+                EnGiant_ChangeToStartOrLoopAnimation(this, GIANT_ANIMATION_FALLING_OVER);
+                break;
+            case 8:
+                EnGiant_ChangeAnimation(this, GIANT_ANIMATION_BIG_CALL_LOOP);
+                break;
+            case 9:
+            case 11:
+                EnGiant_ChangeAnimation(this, GIANT_ANIMATION_IDLE_LOOP);
+                break;
+            case 10:
+                EnGiant_ChangeAnimation(this, GIANT_ANIMATION_SMALL_CALL_LOOP);
+                break;
+        }
+        SkelAnime_FrameUpdateMatrix(&this->skelAnime);
+    }
+}
+
+void EnGiant_PlaySound(EnGiant* this) {
+    if (this->actor.draw != NULL && this->alpha > 0) {
+        if (this->animationId == GIANT_ANIMATION_WALKING_LOOP &&
+            (func_801378B8(&this->skelAnime, 40.0f) || func_801378B8(&this->skelAnime, 100.0f))) {
+            Audio_PlayActorSound2(&this->actor, NA_SE_EV_KYOJIN_WALK);
+        }
+        if (this->animationId == GIANT_ANIMATION_FALLING_OVER && func_801378B8(&this->skelAnime, 40.0f)) {
+            Audio_PlayActorSound2(&this->actor, NA_SE_EV_KYOJIN_VOICE_FAIL);
+        }
+        if (this->sfxId != 0xFFFF &&
+            ((this->animationId == GIANT_ANIMATION_BIG_CALL_START && this->skelAnime.animCurrentFrame >= 18.0f) ||
+             this->animationId == GIANT_ANIMATION_BIG_CALL_LOOP)) {
+            func_800B9010(&this->actor, this->sfxId);
+        }
+        if ((this->animationId == GIANT_ANIMATION_SMALL_CALL_START && this->skelAnime.animCurrentFrame >= 18.0f) ||
+            this->animationId == GIANT_ANIMATION_SMALL_CALL_LOOP) {
+            func_800B9010(&this->actor, NA_SE_EV_KYOJIN_SIGN - SFX_FLAG);
+        }
+    }
+}
+
+void EnGiant_UpdatePosition(EnGiant* this, GlobalContext* globalCtx, u32 actionIndex) {
+    CsCmdActorAction* actorAction = globalCtx->csCtx.npcActions[actionIndex];
+    f32 floatUnk10 = actorAction->unk10;
+    s32 pad[2];
+    f32 floatUnk1C = actorAction->unk1C;
```

Comment:
Would be interesting to check the OoT cs system to see if the struct is the same, but you don't have to do it now.

Reply:
It's might be the same; I actually already checked this. Here's the OoT struct:
```c
typedef struct {
    /* 0x00 */ u16 action; // "dousa"
    /* 0x02 */ u16 startFrame;
    /* 0x04 */ u16 endFrame;
    union {
        /* 0x06 */ Vec3s rot;
        /* 0x06 */ Vec3us urot;
    };
    /* 0x0C */ Vec3i startPos;
    /* 0x18 */ Vec3i endPos;
    /* 0x24 */ Vec3i normal;
} CsCmdActorAction; // size = 0x30
```

The MM struct is the same size, and assuming everything is in the same position, this function is checking `startPos.y` and `endPos.y`, which would make sense. Should I port this definition into MM? Should I only bring over `action`, `startPos`, and `endPos`? Or should I do nothing, since I haven't decompiled any cutscene code and thus have no clue what this is really doing?